articles/build-a-ticketing-app-with-adonisjs-and-vuejs/index.md

```diff
@@ -0,0 +1,1000 @@
+
+const routes = [{
+        path: "/",
+        name: "home",
+        component: Home,
+    },
+    {
+        path: "/login",
+        name: "login",
+        component: Login,
+    },
+    {
+        path: "/register",
+        name: "register",
+        component: Register,
+    },
+
+    // Protected Routes here
+];
+const router = createRouter({
+    history: createWebHistory(),
+    routes
+});
+
+export default router;
+```
+
+In the code above, we created different routes for `login`, `register`, `index` and protected routes such as `admin` and `user` below:
+
+```js
+{
+    path: "/user/:id",
+    name: "user",
+    component: User,
+    meta: {
+        requiresAuth: true
+    },
+    beforeEnter(to, from, next) {
+        if (
+            store.getters["isUser"] &&
+            parseInt(store.state.user.id) === parseInt(to.params.id)
+        ) {
+            next();
+        } else {
+            next({
+                name: "login",
+            });
+        }
+    },
+}, {
+    path: "/admin",
+    name: "admin",
+    component: Admin,
+    meta: {
+        requiresAuth: true
+    },
+    children: [{
+            path: "add",
+            component: Add,
+        },
+        {
+            path: "/",
+            component: AdminHome,
+        },
+    ],
+    beforeEnter(to, from, next) {
+        if (store.getters["isAdmin"]) {
+            next();
+        } else {
+            next({
+                name: "login",
+            });
+        }
+    },
+},
+```
+
+The `beforeEach` route hook checks each route for any authentication metadata and if found, checks if the user navigating that route has logged in or not then redirects.
+
+```js
+router.beforeEach((to, from, next) => {
+    to.matched.some((record) => {
+        console.log(record);
+        return record.meta.requiresAuth;
+    });
+    if (to.matched.some((record) => record.meta.requiresAuth)) {
+        if (!store.state.loggedIn) {
+            next({
+                path: "/login",
+                query: {
+                    redirect: to.fullPath
+                },
+            });
+        } else {
+            next();
+        }
+    } else {
+        next();
+    }
+});
+```
+
+To get the routes to work, we added the new route package and registered this file inside the `main.js` file so we can use it as Vue 3 plugin.
+
+```bash
+npm install vue-router@4
+```
+
+Inside the `main.js`, we added it as below:
+
+```js
+import { createApp } from "vue";
+import router from "./routes";
+//........
+
+createApp(App).use(router).use(store).mount("#app");
+```
+
+#### Set up user authentication
+We will start by creating the login and register views in the `src/views` folder, run the following commands to create these files.
+
+```bash
+touch login.vue
+touch register.vue
+```
+
+Next, we will add the following code to create in the `login.vue` to create the login form and also the login process with error handling.
+
+```html
```

Comment:
here - just wanted to make sure that there is not a missing line of code here @Kaperskyguru 
cc @srishilesh

Reply:
None.. The codes are listed below.

Reply:
Okay - I thought so just wasn't sure since the text block looked empty. Thank you.